Add support for rendering waveform images outside of the view [WIP]

ospr opened the pull request:
I needed support for rending the waveform image outside of the view so I could cache it as well as support for showing a linear scaled waveform. This PR includes those features as well as several bugs that were fixed along side the massive refactor:

**Features Added:**
- Added support for rendering waveform images outside of a view (See `FDWaveformRenderOperation`).
- Added support for rendering linear waveforms.
- Added support for changing `wavesColor` and `progressColor` after waveform was rendered.
- Added support for updating waveform type and color to iOS Example app.

**Bugs Fixed:**
- Fixed bug which could prevent waveform from fitting new view size if rendering was in progress during a view resize.
- Fixed bug which caused `waveformViewDidLoad()` to not be called after the audio file was loaded.
- Fixed bug which caused subsequent waveform renderings for new audioURLs to never complete if there was an error with a previous render.
- Fixed bug which could cause a crash (divide by zero error) if the view's width was 0.

Tested performance using iOS Example's performance testing and confirmed there were no regressions. Tested linear/log and color changing support in added features in iOS Example app.

Note that this also includes the changes made in my previous PR https://github.com/fulldecent/FDWaveformView/pull/79.

fulldecent commented:
Thank you, I am looking into this. Specifically, first I am looking for parts that can be accepted without changing too much.

ospr commented:
Thanks for taking a look. I initially started down the path of breaking the rendering out into its own operation, then came across the other fixes as I was working on them.

The following could probably be broken out of this large pr: 

**Features Added:**
- Added support for changing `wavesColor` and `progressColor` after waveform was rendered.
- Added support for updating color to iOS Example app.

**Bugs Fixed:**
- Fixed bug which caused `waveformViewDidLoad()` to not be called after the audio file was loaded.
- Fixed bug which could cause a crash (divide by zero error) if the view's width was 0.

The other fixes/features are more tightly integrated with the refactoring to the rendering operation. Could probably get away with adding linear support as well without the rendering operation, but the waveform view is already doing a lot and the refactor I think helps distribute the roles a bit better.

fulldecent commented:
I don't want to lose this. And also, I'd like to learn more about all this before #86.

fulldecent commented on Source/FDWaveformView.swift:
```diff
@@ -158,16 +174,18 @@ open class FDWaveformView: UIView {
 
     /// View for rendered waveform
-    fileprivate let imageView: UIImageView = {
+    lazy fileprivate var imageView: UIImageView = {
```
Why var?

ospr replied:
Swift requires lazy properties to be var instead of let and the property was made lazy to allow for initializing the image view with the waves color.

fulldecent commented:
I didn't finish studying this. But it looks like this makes progress on #2.

fulldecent commented:
Ok, I have studied this and gotten the merge to work. Thanks again, this is really a lot of great stuff. There are few remaining issues:

- [ ] Rendering time was increased 4x in my tests -- this is a blocker
  - Perhaps this is because the tight loop calls `process(normalizedSamples: inout [Float])`
  - Is vDSP using the same bucket size as before?
- [ ] Break out code into separate files, as appropriate
- [ ] Consider the exposed public API (use Related Items -> Counterparts -> FDWaveformView.swift), there are several items that should be `internal`

@ospr Do you have any advice on these items or would you like to help with them?

ospr commented:
Hey @fulldecent thanks for taking a look at this. Unfortunately my recently acquired job doesn't allow for contributions to open source, so I'm not sure I'll be able to help much.

Regarding rendering time, that sounds strange to me as during my tests I don't recall finding regressions in rendering time.

fulldecent commented:
Thanks for the feedback. Maybe my computer is the one with the issue. Either way, I'll merge now and if there's a regression that will be a separate issue.
